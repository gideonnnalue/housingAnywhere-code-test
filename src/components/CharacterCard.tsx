import { FC, useState } from "react";
import ChapterModal from "./ChapterModal";
import { ICharacterProps } from "./types";

const CharacterCard: FC<ICharacterProps> = (props): JSX.Element => {
  const [modalOpen, setModalOpen] = useState<boolean>(false);
  const toggleModalState = (e: any) => {
    e.stopPropagation();
    setModalOpen((prev) => !prev);
  };
  const { image, origin, location, name, episode, species, status, gender } =
    props;
  const episodes = episode.reduce((current, next) => {
    const newCurr = [...current];
    newCurr.push(next.data);
    return newCurr;
  }, []);
  return (
    <div className="characterCard">
      <img className="characterCard__img" src={image} alt="character" />
      <div className="characterCard__content">
        <div className="characterCard__top">
          <h3 className="characterCard__name">{name}</h3>
          <p className="characterCard__status">
            {status}
            <span className={status}></span>
          </p>
        </div>
        <div className="characterCard__desc">
          <div className="characterCard__block">
            <div className="characterCard__info">
              <p className="characterCard__info--title">Gender</p>
              <p className="characterCard__info--text">{gender}</p>
            </div>
            <div className="characterCard__info">
              <p className="characterCard__info--title">Specie</p>
              <p className="characterCard__info--text">{species}</p>
            </div>
          </div>
          <h4>ORIGIN INFORMATION</h4>
          <div className="characterCard__block">
            <div className="characterCard__info">
              <p className="characterCard__info--title">NAME</p>
              <p className="characterCard__info--text">{origin.name}</p>
            </div>
            <div className="characterCard__info">
              <p className="characterCard__info--title">DIMENSION</p>
              <p className="characterCard__info--text">
                {origin.dimension ? origin.dimension : "unknown"}
              </p>
            </div>
            <div className="characterCard__info">
              <p className="characterCard__info--title">RESIDENTS</p>
              <p className="characterCard__info--text">
                {origin.residents ? origin.residents.length : 0}
              </p>
            </div>
          </div>

          <h4>LOCATION INFORMATION</h4>
          <div className="characterCard__block">
            <div className="characterCard__info">
              <p className="characterCard__info--title">NAME</p>
              <p className="characterCard__info--text">{location.name}</p>
            </div>
            <div className="characterCard__info">
              <p className="characterCard__info--title">DIMENSION</p>
              <p className="characterCard__info--text">
                {location.dimension ? location.dimension : "unknown"}
              </p>
            </div>
            <div className="characterCard__info">
              <p className="characterCard__info--title">RESIDENTS</p>
              <p className="characterCard__info--text">
                {location.residents ? location.residents.length : 0}
              </p>
            </div>
          </div>

          <div className="characterCard__block">
            <div className="characterCard__info">
              <p className="characterCard__info--title">CHAPTERS</p>
              <div className="characterCard__info--text">
                <p>
                  Character featured in <span>{episode.length}</span> chapters
                </p>
                <button onClick={toggleModalState}>
                  View Chapter information
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      {modalOpen && (
        <ChapterModal episodes={episodes} closeModal={toggleModalState} />
      )}
    </div>
  );
};

export default CharacterCard;
