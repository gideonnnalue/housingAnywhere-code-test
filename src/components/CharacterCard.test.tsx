import { render, cleanup } from "@testing-library/react";

import CharacterCard from "./CharacterCard";
const episodes = [
  {
    air_date: "January 27, 2014",
    characters: [
      "https://rickandmortyapi.com/api/character/1",
      "https://rickandmortyapi.com/api/character/2",
    ],
    created: "2017-11-10T12:56:34.339Z",
    episode: "S01E06",
    id: 6,
    name: "Rick Potion #9",
    url: "https://rickandmortyapi.com/api/episode/6",
  },
  {
    air_date: "January 27, 2014",
    characters: [
      "https://rickandmortyapi.com/api/character/1",
      "https://rickandmortyapi.com/api/character/2",
    ],
    created: "2017-11-10T12:56:34.339Z",
    episode: "S01E06",
    id: 7,
    name: "Rick Potion #9",
    url: "https://rickandmortyapi.com/api/episode/6",
  },
];

const localData = {
  name: "Earth (Replacement Dimension)",
  dimension: "Earth (Replacement Dimension)",
  residents: ["1", "2", "3"],
  url: "https://rickandmortyapi.com/api/episode/6",
  created: null,
  id: null,
  type: null,
};
const data = {
  image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
  origin: localData,
  location: localData,
  name: "Rick Sanchez",
  episode: episodes,
  species: "Human",
  status: "Alive",
  gender: "Male",
};

afterEach(cleanup);

describe("Character card test", () => {
  test("Should take snapshot of card", () => {
    const { asFragment } = render(
      <CharacterCard
        image={data.image}
        origin={data.origin}
        location={data.location}
        name={data.name}
        episode={data.episode}
        species={data.species}
        status={data.status}
        gender={data.status}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
