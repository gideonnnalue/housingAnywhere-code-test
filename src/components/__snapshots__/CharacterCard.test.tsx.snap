// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character card test Should take snapshot of card 1`] = `
<DocumentFragment>
  <div
    class="characterCard"
  >
    <img
      alt="character"
      class="characterCard__img"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
    <div
      class="characterCard__content"
    >
      <div
        class="characterCard__top"
      >
        <h3
          class="characterCard__name"
        >
          Rick Sanchez
        </h3>
        <p
          class="characterCard__status"
        >
          Alive
          <span
            class="Alive"
          />
        </p>
      </div>
      <div
        class="characterCard__desc"
      >
        <div
          class="characterCard__block"
        >
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              Gender
            </p>
            <p
              class="characterCard__info--text"
            >
              Alive
            </p>
          </div>
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              Specie
            </p>
            <p
              class="characterCard__info--text"
            >
              Human
            </p>
          </div>
        </div>
        <h4>
          ORIGIN INFORMATION
        </h4>
        <div
          class="characterCard__block"
        >
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              NAME
            </p>
            <p
              class="characterCard__info--text"
            >
              Earth (Replacement Dimension)
            </p>
          </div>
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              DIMENSION
            </p>
            <p
              class="characterCard__info--text"
            >
              Earth (Replacement Dimension)
            </p>
          </div>
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              RESIDENTS
            </p>
            <p
              class="characterCard__info--text"
            >
              3
            </p>
          </div>
        </div>
        <h4>
          LOCATION INFORMATION
        </h4>
        <div
          class="characterCard__block"
        >
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              NAME
            </p>
            <p
              class="characterCard__info--text"
            >
              Earth (Replacement Dimension)
            </p>
          </div>
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              DIMENSION
            </p>
            <p
              class="characterCard__info--text"
            >
              Earth (Replacement Dimension)
            </p>
          </div>
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              RESIDENTS
            </p>
            <p
              class="characterCard__info--text"
            >
              3
            </p>
          </div>
        </div>
        <div
          class="characterCard__block"
        >
          <div
            class="characterCard__info"
          >
            <p
              class="characterCard__info--title"
            >
              CHAPTERS
            </p>
            <div
              class="characterCard__info--text"
            >
              <p>
                Character featured in 
                <span>
                  2
                </span>
                 chapters
              </p>
              <button>
                View Chapter information
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
